lib/demo/accounts/user.ex

```diff
@@ -0,0 +1,122 @@
+defmodule Demo.Accounts.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  @derive {Inspect, except: [:password]}
+  schema "users" do
+    field :email, :string
+    field :password, :string, virtual: true
+    field :hashed_password, :string
+    field :confirmed_at, :naive_datetime
+
+    timestamps()
+  end
+
+  @doc """
+  A user changeset for registration.
+
+  It is important to validate the length of both e-mail and password.
+  Otherwise databases may truncate the e-mail without warnings, which
+  could lead to unpredictable or insecure behaviour. Long passwords may
+  also be very expensive to hash for certain algorithms.
+  """
+  def registration_changeset(user, attrs) do
+    user
+    |> cast(attrs, [:email, :password])
+    |> validate_email()
+    |> validate_password()
+  end
+
+  defp validate_email(changeset) do
+    changeset
+    |> validate_required([:email])
+    |> validate_format(:email, ~r/^[^\s]+@[^\s]+$/, message: "must have the @ sign and no spaces")
+    |> validate_length(:email, max: 160)
+    |> unsafe_validate_unique(:email, Demo.Repo)
+    |> unique_constraint(:email)
+  end
+
+  defp validate_password(changeset) do
+    changeset
+    |> validate_required([:password])
+    |> validate_length(:password, min: 12, max: 80)
+    # |> validate_format(:password, ~r/[a-z]/, message: "at least one lower case character")
+    # |> validate_format(:password, ~r/[A-Z]/, message: "at least one upper case character")
+    # |> validate_format(:password, ~r/[!?@#$%^&*_0-9]/, message: "at least one digit or punctuation character")
+    |> maybe_hash_password()
```

Comment:
Out of curiosity, should this be wrapped by a `prepare_changes/1` call?  It seems like that would make it so the password is only hashed on save, instead of every time this function is called. I'd think that would provide some performance benefits if this changeset was used in a LiveView where this function was called on every change event. wdyt?

Reply:
@aaronrenner I like that, let's do it!

Reply:
On the other hand, if we are thinking about LiveView, this means that the password will only be validated on submission. Wouldn't `phx-debounce="blur"` be a better call for LiveView?

Reply:
I was talking about the contents of `maybe_hash_password` would run inside a prepare_changes callback (not the full validation).

There are a few ways I could see this being written...

```elixir
def validate_password(changeset)
  changeset
  |> #...
  |> prepare_changes(&maybe_hash_password/1)
end

def maybe_hash_password(changeset) do
  # Stays the same
end
```

or

```elixir
def validate_password(changeset)
  # Stays the same
end

defp maybe_hash_password(changeset) do
  prepare_changes(changeset, fn changeset ->
    if password = get_change(changeset, :password) do
      changeset
      |> put_change(:hashed_password, Bcrypt.hash_pwd_salt(password)
      |> delete_change(:password)
    else
      changeset
    end
  end
end
```

or maybe even

```elixir
def validate_password(changeset)
  changeset
  |> #...
  |> prepare_for_write()
end

defp prepare_for_write(changeset) do
  prepare_changes(changeset, &maybe_hash_password)
end

defp maybe_hash_password(changeset) do
  if password = get_change(changeset, :password) do
    changeset
    |> put_change(:hashed_password, Bcrypt.hash_pwd_salt(password)
    |> delete_change(:password)
  else
    changeset
  end
end
```

Stylistically, I'm not sure which is the preferred pattern, but the idea is that the server would only hash the password right before its sent to the db (and the rest of the validations would run every time).

Reply:
Ah, you are correct.

```
  |> prepare_changes(&maybe_hash_password/1)
```

☝️ this is my vote.